### PR TITLE
[MAINTENANCE] Use datasource config in add_datasource support methods

### PR DESCRIPTION
**Pull request by anthonyburdi**
Changes proposed in this pull request:
- Instead of passing name and config around separately, refactor support methods for context.add_datasource() to use DatasourceConfig objects containing the name param
- Some light refactoring e.g. adding AbstractDataContext._perform_substitutions_on_datasource_config()

### Definition of Done

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 1]
```diff
@@ -491,11 +494,15 @@ def add_datasource(
         else:
             config = kwargs
 
+        datasource_config: DatasourceConfig = datasourceConfigSchema.load(
+            CommentedMap(**config)
+        )
```
Comment (cdkini) [inline great_expectations/data_context/data_context/abstract_data_context.py:499]: Is there a reason we use `CommentedMap` here?

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 2]
```diff
@@ -1679,44 +1692,75 @@ def _build_datasource_from_config(
...
+
+        Args:
+            config: Datasource Config
+
+        Returns:
+            Datasource Config with substitutions performed.
+        """
+        substitutions: dict = self._determine_substitutions()
+
+        substitution_serializer = DictConfigSerializer(schema=datasourceConfigSchema)
```
Comment (cdkini) [inline great_expectations/data_context/data_context/abstract_data_context.py:1710]: When should these values be provided as args (DI) vs being instantiated inline? Just want to get a better mental model around this.

File: great_expectations/data_context/data_context/base_data_context.py [hunk 3]
```diff
@@ -2898,27 +2899,23 @@ def _test_instantiation_of_misc_class_from_yaml_config(
 
         return instantiated_class, usage_stats_event_payload
 
-    def _instantiate_datasource_from_config_and_update_project_config(
+    def _instantiate_datasource_from_config_and_update_project_config(  # type: ignore
```
Comment (cdkini) [inline great_expectations/data_context/data_context/base_data_context.py]: any specific warning we're ignoring here? I'd like to use the `[specific-warning]` syntax as much as possible

File: great_expectations/data_context/data_context/abstract_data_context.py (hunk 1 above)
Comment (anthonyburdi) [reply great_expectations/data_context/data_context/abstract_data_context.py:499]: I believe this supports comments in yaml, I left it in to support existing functionality

File: great_expectations/data_context/data_context/abstract_data_context.py (hunk 2 above)
Comment (anthonyburdi) [reply great_expectations/data_context/data_context/abstract_data_context.py:1710]: The way I'm thinking about it is that if I can see a reason we would want to use a different serializer I provide as an arg, if not then inline.

File: great_expectations/data_context/data_context/base_data_context.py (hunk 3 above)
Comment (anthonyburdi) [reply great_expectations/data_context/data_context/base_data_context.py]: Good catch, missed this one. Fixing.

File: great_expectations/data_context/data_context/abstract_data_context.py (hunk 1 above)
Comment (cdkini) [reply great_expectations/data_context/data_context/abstract_data_context.py:499]: works for me! I'd rather not expose `ruamel` internals outside of `YAMLHandler` but this isn't in scope for your PR

File: great_expectations/data_context/data_context/abstract_data_context.py (hunk 2 above)
Comment (cdkini) [reply great_expectations/data_context/data_context/abstract_data_context.py:1710]: works for me!

**Review by cdkini — APPROVED** [review]

File: great_expectations/data_context/data_context/base_data_context.py (hunk 3 above)
Comment (anthonyburdi) [reply great_expectations/data_context/data_context/base_data_context.py]: Fixed!

**Pull request merged**